packages/snaps-execution-environments/src/common/BaseSnapExecutor.ts

```diff
@@ -362,17 +372,41 @@ export class BaseSnapExecutor {
   }
 
   /**
-   * Instantiates a snap provider object (i.e. `globalThis.wallet`).
+   * Instantiates a snap API object (i.e. `globalThis.snap`).
    *
+   * @param provider - A StreamProvider connected to MetaMask.
    * @returns The snap provider object.
    */
-  private createSnapProvider(): SnapProvider {
-    const provider = new MetaMaskInpageProvider(this.rpcStream, {
-      shouldSendMetadata: false,
-    });
+  private createSnapAPI(provider: StreamProvider): SnapAPI {
+    const originalRequest = provider.request;
+
+    const request = async (args: RequestArguments) => {
+      assert(
+        args.method.startsWith('wallet_') || args.method.startsWith('snap_'),
+        'Snap API only allows RPC methods wallet_* and snap_*',
+      );
+      this.notify({ method: 'OutboundRequest' });
+      try {
+        return await withTeardown(originalRequest(args), this as any);
+      } finally {
+        this.notify({ method: 'OutboundResponse' });
+      }
+    };
+
+    return { request };
+  }
+
+  /**
+   * Instantiates an eip1193 provider object (i.e. `globalThis.ethereum`).
+   *
+   * @param provider - A StreamProvider connected to MetaMask.
+   * @returns The eip1193 provider object.
+   */
+  private createEIP1193Provider(provider: StreamProvider): StreamProvider {
     const originalRequest = provider.request;
 
     provider.request = async (args) => {
+      assert(!args.method.startsWith('snap_'), 'RPC method not allowed');
```

Comment:
Maybe just throw a "method not found" JSON-RPC error.

Reply:
`assert(!args.method.startsWith('snap_'), ethErrors.rpc.methodNotFound());` ?

Reply:
Yep